File: actito_inbox/android/src/main/kotlin/com/actito/inbox/flutter/ActitoInboxPlugin.kt
package com.actito.inbox.flutter

import android.os.Handler
import android.os.Looper
import androidx.annotation.NonNull
import androidx.lifecycle.Observer
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.JSONMethodCodec
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONObject
import com.actito.Actito
import com.actito.ActitoCallback
import com.actito.inbox.flutter.events.ActitoEvent
import com.actito.inbox.flutter.events.ActitoEventManager
import com.actito.inbox.ktx.inbox
import com.actito.inbox.models.ActitoInboxItem
import com.actito.models.ActitoNotification
import java.util.*

class ActitoInboxPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel

    private val itemsObserver = Observer<SortedSet<ActitoInboxItem>> { items ->
        if (items == null) return@Observer

        ActitoEventManager.send(
            ActitoEvent.InboxUpdated(items)
        )
    }

    private val badgeObserver = Observer<Int> { badge ->
        if (badge == null) return@Observer

        ActitoEventManager.send(
            ActitoEvent.BadgeUpdated(badge)
        )
    }

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        ActitoEventManager.register(binding.binaryMessenger)

        channel = MethodChannel(
            binding.binaryMessenger,
            "com.actito.inbox.flutter/actito_inbox",
            JSONMethodCodec.INSTANCE
        )
        channel.setMethodCallHandler(this)

        Actito.inbox().observableItems.observeForever(itemsObserver)
        Actito.inbox().observableBadge.observeForever(badgeObserver)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        Actito.inbox().observableItems.removeObserver(itemsObserver)
        Actito.inbox().observableBadge.removeObserver(badgeObserver)

        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "getItems" -> getItems(call, result)
            "getBadge" -> getBadge(call, result)
            "refresh" -> refresh(call, result)
            "open" -> open(call, result)
            "markAsRead" -> markAsRead(call, result)
            "markAllAsRead" -> markAllAsRead(call, result)
            "remove" -> remove(call, result)
            "clear" -> clear(call, result)
            else -> result.notImplemented()
        }
    }

    private fun getItems(@Suppress("UNUSED_PARAMETER") call: MethodCall, result: Result) {
        onMainThread {
            result.success(
                Actito.inbox().items.map { it.toJson() }
            )
        }
    }

    private fun getBadge(@Suppress("UNUSED_PARAMETER") call: MethodCall, result: Result) {
        onMainThread {
            result.success(Actito.inbox().badge)
        }
    }

    private fun refresh(@Suppress("UNUSED_PARAMETER") call: MethodCall, result: Result) {
        Actito.inbox().refresh()
        result.success(null)
    }

    private fun open(@Suppress("UNUSED_PARAMETER") call: MethodCall, pluginResult: Result) {
        val arguments = call.arguments<JSONObject>() ?: return onMainThread {
            pluginResult.error(ACTITO_ERROR, "Invalid request arguments.", null)
        }

        val item = ActitoInboxItem.fromJson(arguments)

        Actito.inbox().open(item, object : ActitoCallback<ActitoNotification> {
            override fun onSuccess(result: ActitoNotification) {
                onMainThread {
                    pluginResult.success(result.toJson())
                }
            }

            override fun onFailure(e: Exception) {
                onMainThread {
                    pluginResult.error(ACTITO_ERROR, e.localizedMessage, null)
                }
            }
        })
    }

    private fun markAsRead(@Suppress("UNUSED_PARAMETER") call: MethodCall, pluginResult: Result) {
        val arguments = call.arguments<JSONObject>() ?: return onMainThread {
            pluginResult.error(ACTITO_ERROR, "Invalid request arguments.", null)
        }

        val item = ActitoInboxItem.fromJson(arguments)

        Actito.inbox().markAsRead(item, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                pluginResult.success(null)
            }

            override fun onFailure(e: Exception) {
                pluginResult.error(ACTITO_ERROR, e.localizedMessage, null)
            }
        })
    }

    private fun markAllAsRead(@Suppress("UNUSED_PARAMETER") call: MethodCall, pluginResult: Result) {
        Actito.inbox().markAllAsRead(object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                pluginResult.success(null)
            }

            override fun onFailure(e: Exception) {
                pluginResult.error(ACTITO_ERROR, e.localizedMessage, null)
            }
        })
    }

    private fun remove(@Suppress("UNUSED_PARAMETER") call: MethodCall, pluginResult: Result) {
        val arguments = call.arguments<JSONObject>() ?: return onMainThread {
            pluginResult.error(ACTITO_ERROR, "Invalid request arguments.", null)
        }

        val item = ActitoInboxItem.fromJson(arguments)

        Actito.inbox().remove(item, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                pluginResult.success(null)
            }

            override fun onFailure(e: Exception) {
                pluginResult.error(ACTITO_ERROR, e.localizedMessage, null)
            }
        })
    }

    private fun clear(@Suppress("UNUSED_PARAMETER") call: MethodCall, pluginResult: Result) {
        Actito.inbox().clear(object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                pluginResult.success(null)
            }

            override fun onFailure(e: Exception) {
                pluginResult.error(ACTITO_ERROR, e.localizedMessage, null)
            }
        })
    }

    companion object {
        internal const val ACTITO_ERROR = "actito_error"

        internal fun onMainThread(action: () -> Unit) {
            Handler(Looper.getMainLooper()).post { action() }
        }
    }
}

File: actito_inbox/ios/actito_inbox/Sources/actito_inbox/ActitoInboxPlugin.swift
import ActitoInboxKit
import ActitoKit
import Flutter
import UIKit

private let DEFAULT_ERROR_CODE = "actito_error"

public class ActitoInboxPlugin: NSObject, FlutterPlugin {
    static let instance = ActitoInboxPlugin()

    public static func register(with registrar: FlutterPluginRegistrar) {
        instance.register(with: registrar)
    }

    private let events = ActitoInboxPluginEvents(packageId: "com.actito.inbox.flutter")

    private func register(with registrar: FlutterPluginRegistrar) {
        // Delegate
        onMainThreadIsolated {
            Actito.shared.inbox().delegate = self
        }

        // Events
        events.setup(registrar: registrar)

        // Communication channel
        let channel = FlutterMethodChannel(
            name: "com.actito.inbox.flutter/actito_inbox",
            binaryMessenger: registrar.messenger(),
            codec: FlutterJSONMethodCodec.sharedInstance()
        )

        channel.setMethodCallHandler { (call, result) in
            switch call.method {
            case "getItems": self.getItems(call, result)
            case "getBadge": self.getBadge(call, result)
            case "refresh": self.refresh(call, result)
            case "open": self.open(call, result)
            case "markAsRead": self.markAsRead(call, result)
            case "markAllAsRead": self.markAllAsRead(call, result)
            case "remove": self.remove(call, result)
            case "clear": self.clear(call, result)

            // Unhandled
            default: result(FlutterMethodNotImplemented)
            }
        }
    }

    private func getItems(_ call: FlutterMethodCall, _ response: @escaping FlutterResult) {
        DispatchQueue.main.async {
            do {
                let items = try Actito.shared.inbox().items.map { item in
                    try item.toJson()
                }

                response(items)
            } catch {
                response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            }
        }
    }

    private func getBadge(_ call: FlutterMethodCall, _ response: @escaping FlutterResult) {
        DispatchQueue.main.async {
            response(Actito.shared.inbox().badge)
        }
    }

    private func refresh(_ call: FlutterMethodCall, _ response: @escaping FlutterResult) {
        DispatchQueue.main.async {
            Actito.shared.inbox().refresh { result in
                switch result {
                case .success:
                    response(nil)
                case let .failure(error):
                    response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func open(_ call: FlutterMethodCall, _ response: @escaping FlutterResult) {
        let item: ActitoInboxItem

        do {
            let json = call.arguments as! [String: Any]
            item = try ActitoInboxItem.fromJson(json: json)
        } catch {
            response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            return
        }

        DispatchQueue.main.async {
            Actito.shared.inbox().open(item) { result in
                switch result {
                case let .success(notification):
                    do {
                        let json = try notification.toJson()
                        response(json)
                    } catch {
                        response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
                    }
                case let .failure(error):
                    response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func markAsRead(_ call: FlutterMethodCall, _ response: @escaping FlutterResult) {
        let item: ActitoInboxItem

        do {
            let json = call.arguments as! [String: Any]
            item = try ActitoInboxItem.fromJson(json: json)
        } catch {
            response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            return
        }

        DispatchQueue.main.async {
            Actito.shared.inbox().markAsRead(item) { result in
                switch result {
                case .success:
                    response(nil)
                case let .failure(error):
                    response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func markAllAsRead(_ call: FlutterMethodCall, _ response: @escaping FlutterResult) {
        DispatchQueue.main.async {
            Actito.shared.inbox().markAllAsRead { result in
                switch result {
                case .success:
                    response(nil)
                case let .failure(error):
                    response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func remove(_ call: FlutterMethodCall, _ response: @escaping FlutterResult) {
        let item: ActitoInboxItem

        do {
            let json = call.arguments as! [String: Any]
            item = try ActitoInboxItem.fromJson(json: json)
        } catch {
            response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            return
        }

        DispatchQueue.main.async {
            Actito.shared.inbox().remove(item) { result in
                switch result {
                case .success:
                    response(nil)
                case let .failure(error):
                    response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func clear(_ call: FlutterMethodCall, _ response: @escaping FlutterResult) {
        DispatchQueue.main.async {
            Actito.shared.inbox().clear { result in
                switch result {
                case .success:
                    response(nil)
                case let .failure(error):
                    response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
                }
            }
        }
    }
}

extension ActitoInboxPlugin: ActitoInboxDelegate {
    public func actito(_ actitoInbox: ActitoInbox, didUpdateInbox items: [ActitoInboxItem]) {
        events.emit(
            ActitoInboxPluginEvents.OnInboxUpdated(items: items)
        )
    }

    public func actito(_ actitoInbox: ActitoInbox, didUpdateBadge badge: Int) {
        events.emit(
            ActitoInboxPluginEvents.OnBadgeUpdated(badge: badge)
        )
    }
}

internal func onMainThreadIsolated<T>(_ block: @MainActor @escaping () -> T) -> T {
    if Thread.isMainThread {
        return MainActor.assumeIsolated {
            block()
        }
    } else {
        let group = DispatchGroup()
        var result: T! = nil

        group.enter()

        DispatchQueue.main.async {
            result = block()
            group.leave()
        }

        group.wait()
        return result
    }
}
